src/java/voldemort/store/socket/clientrequest/ClientRequestExecutor.java

```diff
@@ -235,6 +257,38 @@ protected void read(SelectionKey selectionKey) throws IOException {
     }
 
     @Override
+    protected void reportException(IOException e) {
+        clientRequest.reportException(e);
+    }
+
+    @Override
+    protected void connect(SelectionKey selectionKey) throws IOException {
```

Comment:
There are 2 main issues with the Selector today. One is Connect blocks. The second is Protocol Negotiation blocks as well. https://github.com/voldemort/voldemort/blob/master/src/java/voldemort/store/socket/clientrequest/ClientRequestExecutorFactory.java#L209

Selector should be running in a tight loop reading/writing from sockets. If Selector blocks during connect/protocol negotiation, though the server could have responded the client requests can't receive them as Selector is blocked. This is the root cause of the Search issue, we investigated for more than 2-3 years. On Search, when a node was dying it accepted connect, but never sent the ACK, so the selector blocked and all the other requests queued on the same selector timed out. So we were marking random nodes down. 

On experimental runs in our staging environment for zoned clusters the patch, keeps the 95th percentile at 3 ms(min 2, max 5) and 99th at 5ms.  Unpatched current clients have 95th at 4ms( min 2, max 100) and 99th 10ms( min 4, max 700). This patch lets you have the predictability and after this fix Selector will never block.

So on a high level, this lets you achieve 2 things.
1) Handle the failure scenarios well on a non-zoned cluster. This was always a problem today for us, if a node goes down most of our clients need bouncing.
2) On a Zoned cluster, this gives you a great latency improvements.